### PR TITLE
Feat/handle cycles round 2

### PR DESCRIPTION
**Pull request by aviadhahami**
- [X] Ready for review
- [X] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Added more cycle-specific tests to verify expected-shape changes, cycle handling where the graph is heavily cycled (surreal project), and few comments for the next one who drops in.

I'll follow this by a pr w/ and Object implementation (rather than `Set`) for performance improvement (will be benchmarked vs the current solution first)

### CODE REVIEW

File: src/core/builder.ts [hunk 1]
```diff
@@ -46,7 +46,7 @@ class DepGraphBuilder {
     this._pkgManager = pkgManager;
   }
 
-  // TODO: this can create disconnected nodes
+  // TODO: this can create disconnected nodes -- and memory leaks!
```
Comment (michael-go) [inline src/core/builder.ts:49]: how?

File: src/core/dep-graph.ts [hunk 2]
```diff
@@ -325,7 +325,7 @@ class DepGraphImpl implements types.DepGraphInternal {
       .map((parentId) => {
         const pathsFromParent = this.pathsFromNodeToRoot(
           parentId,
-          new Set(seenOnPath).add(nodeId),
+          new Set(seenOnPath).add(nodeId), // Max size is O(log(n)) [tree height] where n is the amount of nodes in the tree
```
Comment (michael-go) [inline src/core/dep-graph.ts:328]: maybe "tree depth" is a more common term?

Comment (aviadhahami) [reply src/core/dep-graph.ts:328]: I guess it depends on the school :) 
will update it

File: src/core/builder.ts (hunk 1 above)
Comment (aviadhahami) [reply src/core/builder.ts]: added this note prior to seeing the underlying `graph` object

File: src/legacy/index.ts [hunk 3]
```diff
@@ -314,8 +323,10 @@ async function buildSubtree(
         addLabel(depTree, 'pruned', 'true');
       }
       return depTree;
+    }else {
```
Comment (michael-go) [inline src/legacy/index.ts:326]: missing a white space. did you `npm run format` ?

Comment (aviadhahami) [reply src/legacy/index.ts]: :O 
fixed!

File: test/core/direct-deps-leading-to.test.ts [hunk 4]
```diff
@@ -41,13 +41,28 @@ describe('directDepsLeadingTo', () => {
...
+    test('it works with a simple cyclic dep-graph', () => {
+      const cyclic = depGraphLib.createFromJSON(
+        helpers.loadFixture('cyclic-dep-graph.json'),
+      );
+      const pkg = { name: 'baz', version: '4' };
+      const expected = [{ name: 'foo', version: '2' }];
+
+      expect(cyclic.directDepsLeadingTo(pkg)).toEqual(expected);
+    });
+    test('it works with a surreal cyclic dep-graph', () => {
```
Comment (michael-go) [inline test/core/direct-deps-leading-to.test.ts:54]: can you please help my OCD and add a whitespace between tests? 🙇

File: src/legacy/index.ts [hunk 5]
```diff
@@ -314,8 +323,10 @@ async function buildSubtree(
         addLabel(depTree, 'pruned', 'true');
       }
       return depTree;
+    } else {
+      // No double insertions, save some time
```
Comment (michael-go) [inline src/legacy/index.ts:327]: this comment is clear as part of this PR, but can become confusing by itself, so 👍 on the change itself, but maybe remove the comment

File: test/core/direct-deps-leading-to.test.ts (hunk 4 above)
Comment (aviadhahami) [reply test/core/direct-deps-leading-to.test.ts:54]: yes yes :D

**Pull request merged**